cloudinit/sources/helpers/azure.py

```diff
@@ -177,6 +185,46 @@ def report_diagnostic_event(str):
     return evt
 
 
+def report_compressed_event(event_name, event_content):
+    """Report a compressed event"""
+    compressed_data = base64.encodebytes(zlib.compress(event_content))
+    event_data = {"encoding": "gz+b64",
+                  "data": compressed_data.decode('ascii')}
+    evt = events.ReportingEvent(
+        COMPRESSED_EVENT_TYPE, event_name,
+        json.dumps(event_data),
+        events.DEFAULT_EVENT_ORIGIN)
+    events.report_event(evt, excluded_handlers={"logging"})
+
+    # return the event for unit testing purpose
+    return evt
+
+
+@azure_ds_telemetry_reporter
+def push_log_to_kvp(file_name=CFG_BUILTIN['def_log_file']):
+    """Push a portion of cloud-init.log file or the whole file to KVP
+    based on the file size.
+    When called more than once, the function continues pushing the log file
+    from where it left off last time it was called."""
+    global last_log_byte_pushed_to_kvp_index
+    LOG.debug("Dumping cloud-init.log file to KVP")
+
+    try:
+        with open(file_name, "rb") as f:
+            f.seek(0, os.SEEK_END)
+            seek_index = max(f.tell() - MAX_LOG_TO_KVP_LENGTH,
+                             last_log_byte_pushed_to_kvp_index)
+            report_diagnostic_event(
+                "Dumping {} bytes of cloud-init.log file to KVP".format(
+                    f.tell() - seek_index))
+            f.seek(seek_index, os.SEEK_SET)
+            report_compressed_event("cloud-init.log", f.read())
+            last_log_byte_pushed_to_kvp_index = f.tell()
```

Comment:
I think we  actually might want to persist this somewhere outside the global value as cloud-init python process is restarted across every boot stage (init-local, vs init(network) vs comfig:modules etc) and across reboots. 
It looks like most  kvp compressed event pushes occur in "init(network)" after each reboot. If you don't intend to resending the entire cloud-init.log across boots as it grows infinitely, you'd probably want to persist a marker file somewhere in /var/lib/cloud/data/azure_something and read that as your offset.